### PR TITLE
[determinism] Add segment reduction op exceptions for GPU determinism

### PR DESCRIPTION
**Pull request by duncanriach**
## High-Level Summary

This current PR adds and tests the following functionality:

When the environment variable `TF_DETERMINISTIC_OPS` is set to `"true"` or `"1"`, an attempt to run the following ops on a GPU will throw `tf.errors.UnimplementedError` (with an understandable message) when `data` is a floating-point type, including complex types (if supported).

`tf.math.segment_prod`
`tf.math.segment_sum`
`tf.math.unsorted_segment_mean`
`tf.math.unsorted_segment_sqrt_n`
`tf.math.unsorted_segment_prod`
`tf.math.unsorted_segment_sum`

`tf.convert_to_tensor`, when running on a GPU and when `value` is of type `tf.IndexedSlides` will also throw `tf.errors.UnimplementedError` (because it uses `tf.math.unsorted_segment_sum`). This is confirmed by a test included in this PR.

The output of `tf.gather`'s backprop is of type `tf.IndexedSlices`. Therefore, when running on a GPU, if the output of `tf.gather`'s backprop is passed to `tf.convert_to_tensor`, such as when updating a word embedding, then that will also cause `tf.errors.UnimplementedError` to be thrown. This is confirmed by a test included in this PR.

Please see _RFC: Enhancing determinism in TF_ (being added via tensorflow/community PR [346](https://github.com/tensorflow/community/pull/346)).

### Sorted Segment Mean

`tf.math.segment_mean` is currently implemented on the CPU only, operates deterministically, and will not throw `tf.errors.UnimplementedError`. The tests included in the current PR confirm that it does not throw an exception.

### Unsorted Segment Mean and Square Root

`tf.math.unsorted_segment_mean` and `tf.math_unsorted_segment_sqrt_n` are both currently implemented on top of `tf.math.unsorted_segment_sum`. Running these two ops on the GPU, with floating-point-based data types, will cause `tf.errors.UnimplementedError` to be thrown as if by `tf.math.unsorted_segment_sum`.

### Min and Max Segment Reductions

The following ops will not throw `tf.errors.UnimplementedError`, even for floating-point-based types, because min and max operations are associative (ignoring potential non-commutativity and non-associativity of math corner cases). That they do not throw exceptions when run on GPU is confirmed by the tests included in this current PR.

  * `tf.math.segment_max`
  * `tf.math.segment_min`
  * `tf.math.unsorted_segment_max`
  * `tf.math.unsorted_segment_min`

### Data Types

Where GPU implementations of the sorted segment (non-min/max) reduction ops exist, they support the following data types: `tf.float16`, `tf.float32`, and `tf.float64`. The tests confirm that `tf.errors.UnimplementedError` is thrown for all of these, regardless of the integer index data type.

Where GPU implementations of the unsorted segment (non-min/max) reduction ops exist, they support the following data types: `tf.float16`, `tf.float32`, `tf.float64`, and `tf.int32` (except `tf.math.unsorted_segment_sqrt_n`, which does not support `tf.int32`).  The tests confirm that `tf.errors.UnimplementedError` is thrown for all ops on all the floating-point types and not for any of the ops with `int32`, regardless of the integer index data type.

The GPU implementation of `tf.math_unsorted_segment_sum` (and therefore also `tf.math.unsorted_segment_mean` and `tf.math_unsorted_segment_sqrt_n`) supports `tf.complex64` and `tf.complex128`. The tests confirm that `tf.errors.UnimplementedError` is thrown for these ops and datatypes, regardless of the integer index data type.

### Graph Mode vs Eager Mode Testing

All the tests run in eager mode and some tests run in graph mode as well. I was not able to get all the test to work in graph mode, for reasons that seemed unrelated to the actual intentions of the tests. I didn't want to spend more time on it, and I felt that enough of the exception throwing had been shown to translate, as expected, to graph mode.

### XLA

The tests that run in graph mode would also end up getting run with the XLA auto-jit enabled. This would cause tests to fail where the expected exceptions were not generated by the XLA implementations of the ops. For this reason, in this PR, XLA auto-jit is disabled. I don't know if the XLA implementations of these ops are deterministic, but I imagine that they probably would be. I propose that when we implement deterministic versions of these ops, the determinism tests (that actually check determinism) should also be run on the XLA functionality.

### Dense Image Warp

`tfa.image.dense_image_warp`'s backprop to `image` (but not `flow`) is nondeterministic on a GPU because [it uses `tf.gather`](https://github.com/tensorflow/addons/blob/d26e2ed5f68092aed57016a7005ce534b1be3dce/tensorflow_addons/image/dense_image_warp.py#L171). Therefore, after this PR, enabling determinism should cause `tfa.image.dense_image_warp` to throw an exception up from `tf.math.unsorted_segment_sum` if its backprop is run on a GPU. This behavior is not confirmed by the tests included with this current PR.

### CODE REVIEW

**Comment by duncanriach** [comment]
@sanjoy @reedwm : I'm hoping that this can get into TF 2.5

File: tensorflow/core/kernels/segment_reduction_ops.h [hunk 1]
```diff
@@ -76,6 +79,7 @@ struct AtomicSumOpGpu {
                                                         const T& value) {
     GpuAtomicAdd(dest, value);
   }
+  const bool deterministic_for_float = false;
```
Comment (sanjoy) [inline tensorflow/core/kernels/segment_reduction_ops.h:82]: This should say `is_associative` since individual floating point additions are deterministic.  And I'd suggest writing this as:

```
constexpr bool is_associative = std::is_floating_point<T>::value;
```

so that the "caller" does not have to check `T` again.

Same for the other instances.

File: tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc [hunk 2]
```diff
@@ -126,6 +127,29 @@ __global__ void UnsortedSegmentCustomKernel(
...
+bool RequireDeterminism() {
+  static bool require_determinism = [] {
+    bool deterministic_ops = false;
+    TF_CHECK_OK(tensorflow::ReadBoolFromEnvVar("TF_DETERMINISTIC_OPS",
+                                               /*default_val=*/false,
+                                               &deterministic_ops));
+    return deterministic_ops;
+  }();
+  return require_determinism;
+}
```
Comment (sanjoy) [inline tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:139]: Should this live in a single utility somewhere?

File: tensorflow/core/kernels/segment_reduction_ops_impl_1.cc [hunk 3]
```diff
@@ -125,21 +125,26 @@ REGISTER_COMPLEX_CPU_KERNELS_ALL(complex128);
...
   REGISTER_GPU_KERNEL_SORTEDSEGMENT(                                      \
       "SegmentMin", type, index_type, functor::Highest<type>,             \
-      functor::NonAtomicMinOpGpu<type>, functor::AtomicMinOpGpu<type>);   \
+      functor::NonAtomicMinOpGpu<type>, functor::AtomicMinOpGpu<type>,    \
+      /*deterministic_for_float=*/true);                                  \
   REGISTER_GPU_KERNEL_SORTEDSEGMENT(                                      \
       "SegmentMax", type, index_type, functor::Lowest<type>,              \
-      functor::NonAtomicMaxOpGpu<type>, functor::AtomicMaxOpGpu<type>);
+      functor::NonAtomicMaxOpGpu<type>, functor::AtomicMaxOpGpu<type>,    \
+      /*deterministic_for_float=*/true);
```
Comment (sanjoy) [inline tensorflow/core/kernels/segment_reduction_ops_impl_1.cc:147]: Can the `OpKernel`s infer `deterministic_for_float` from the `AtomicFooOpGpu` functors?

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/segment_reduction_ops.h (hunk 1 above)
Comment (duncanriach) [reply tensorflow/core/kernels/segment_reduction_ops.h:82]: Very nice. Will do.

File: tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc (hunk 2 above)
Comment (duncanriach) [reply tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:139]: Probably. I've tried moving it up into a utility in previous PRs but I believe it wasn't accepted because we didn't want to institutionalize something that was not intended to the long-term correct way of doing it (i.e. we'll eventually move to config). I can (re-)scout for a location, move it up, and share it with other code.

File: tensorflow/core/kernels/segment_reduction_ops_impl_1.cc (hunk 3 above)
Comment (duncanriach) [reply tensorflow/core/kernels/segment_reduction_ops_impl_1.cc:147]: I wish they could do. Without completely re-writing a lot of this code, I couldn't figure out a way of doing this the same way for the unsorted ops and for the sorted ops because the code is structured differently for the two sets of ops. For the sorted ops, the reduction functor types bypass the GPU-specific code and appear in non-device-specific code. If I moved the exception-testing into that code, I would need to condition it on device. This repetition-inducing asymmetry has been bothering me too. Let me take a look at this again and see if I can clean it up using a different approach.

Comment (duncanriach) [reply tensorflow/core/kernels/segment_reduction_ops_impl_1.cc:147]: It looks like I misunderstood something about the structure of the code. I think this can actually be cleaned up pretty easily.

Comment (duncanriach) [reply tensorflow/core/kernels/segment_reduction_ops_impl_1.cc:147]: Okay, I'm back to understanding the actual problem that the submitted code solved. The place where we would ideally throw the exception for the sorted segment reduction ops (in order to use the `is_associative` state from `AtomicFooOpGpu`) is running asynchronous compute but doesn't have access to the appropriate `DoneCallBack` in order to call `OP_REQUIRES_ASYNC`. Trying to solve this another way ...

File: tensorflow/core/kernels/segment_reduction_ops.h (hunk 1 above)
Comment (duncanriach) [reply tensorflow/core/kernels/segment_reduction_ops.h]: I made this into a `static constexpr` (class parameter). `constexpr` is not allowed on a non-static member of a class presumably because it has to be assigned (even if const) when the class is instantiated (at runtime and not during compilation). This change, which I think is what we were looking for, may facilitate more compile-time elimination of of some of the logic used to decide branching.

File: tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc (hunk 2 above)
Comment (duncanriach) [reply tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:140]: I added a TODO comment to this. How about I do this (lower-priority) factoring later and in a separate PR? It's going to potentially have a wide span in the TensorFlow source code. This function is currently replicated in several places.

File: tensorflow/core/kernels/segment_reduction_ops_impl_1.cc (hunk 3 above)
Comment (duncanriach) [reply tensorflow/core/kernels/segment_reduction_ops_impl_1.cc]: I think this is good enough now. Please let me know if you can see a cleaner way of doing it.

File: tensorflow/core/kernels/segment_reduction_ops.h (hunk 1 above)
Comment (duncanriach) [reply tensorflow/core/kernels/segment_reduction_ops.h]: Also, I used `std::is_integral<T>::value` because `!std::is_floating_point<T>::value` will not include the complex types. I'm also more confident that all of the integral types are associative than I am confident that all types other than the floating point types (including the complex types) are associative.

**Review by sanjoy — APPROVED** [review]

**Comment by duncanriach** [comment]
@gbaned: please `kokoro:force-run`; I fixed an Ubuntu Sanity error with the most recent commit.

**Comment by duncanriach** [comment]
@sanjoy: I have added a lost of additional information to the [original comment](https://github.com/tensorflow/tensorflow/pull/47772#issue-592163042). My intention is to more thoroughly document this PR both for myself and for anyone else who reviews it in the future. Please take look at it.

**Review by sanjoy — APPROVED** [review]

**Comment by duncanriach** [comment]
Note that non-sparse GPU-deterministic segment reduction ops, both sorted and unsorted, will be added by TF [51392](https://github.com/tensorflow/tensorflow/pull/51392), probably then appearing in release 2.7. The aforementioned PR also adds a (deterministic) GPU implementation of `tf.math.segment_mean` (which hitherto was implemented on CPU only).
